components/qtl-data-view.js

```diff
@@ -1,28 +1,136 @@
 import { DataAreaTitle, DataPanel } from "./data-area";
 import EQTLDataTable from "./eqtl-table";
 
+// To dynamically load component QTLChart on the client side,
+// use the ssr option to disable server-rendering since QTLChart relies on browser APIs like window.
+const QTLChart = dynamic(() => import("./qtl-chart"), {
+  ssr: false,
+});
+
+/**
+ *  Display selected filters and allow user to de-select them.
+ */
+function Selections({ filters, clearFilterFunc }) {
+  return (
+    <div>
+      <span>Selected filters:</span>
+      {filters.map((filter) => {
+        return (
+          <button
+            key={`${filter}-selected-filter`}
+            onClick={() => clearFilterFunc(filter)}
+            className="flex text-data-label text-sm"
+          >
+            <XCircleIcon className="h-5" />
+            {filter}
+          </button>
+        );
+      })}
+    </div>
+  );
+}
+
+Selections.propTypes = {
+  filters: PropTypes.array.isRequired,
+  clearFilterFunc: PropTypes.func.isRequired,
+};
+
 /**
  * This is the view for display QTL data for a variant.
  */
-export function QTLDataView({ data }) {
+export function QTLDataView({ data, assembly }) {
   const [showQtlData, setShowQtlData] = useState(false);
   const router = useRouter();
   useEffect(() => {
     const isQTL = router.asPath.endsWith(`#!qtl`);
     setShowQtlData(isQTL);
   }, [router]);
-  const eQTLData = data.filter((d) => d.method === "eQTLs");
-  const caQTLData = data.filter((d) => d.method === "caQTLs");
+  const [organFilters, setOrganFilters] = useState([]);
+  const filteredData = getFilteredQtlData(data, organFilters);
+  const eQTLData = filteredData.filter((d) => d.method === "eQTLs");
+  const caQTLData = filteredData.filter((d) => d.method === "caQTLs");
+
+  function handleClickOrgan(organ, organList, enabledOrganList) {
+    let filters = [...organFilters];
+    if (organList && enabledOrganList) {
+      if (organList.includes(organ) && enabledOrganList.includes(organ)) {
+        if (filters.includes(organ)) {
+          if (organ in ASSOCIATED_ORGAN_MAP) {
+            ASSOCIATED_ORGAN_MAP[organ].forEach((associatedOrgan) => {
+              if (enabledOrganList.includes(associatedOrgan)) {
+                filters = _.without(filters, associatedOrgan);
+              }
+            });
+          }
+          filters = _.without(filters, organ);
+        } else {
+          filters.push(organ);
+          if (organ in ASSOCIATED_ORGAN_MAP) {
+            ASSOCIATED_ORGAN_MAP[organ].forEach((associatedOrgan) => {
+              if (
+                enabledOrganList.includes(associatedOrgan) &&
+                !filters.includes(associatedOrgan)
+              ) {
+                filters.push(associatedOrgan);
+              }
+            });
+          }
+        }
+        setOrganFilters(filters);
+      }
+    } else {
+      filters = _.without(filters, organ);
+      setOrganFilters(filters);
+    }
+  }
```

Comment:
This is kind of a large function to have the JS interpreter parse every single time `QTLDataView` renders. I think I would make an external function that handles generating the `filters` variable using `organ`, `organList`, `enabledOrganList`, and `organFilters`, call that function from within `handleClickOrgan` which would then be very small, and it pass the `filters` result to `setOrganFilters`. I’ll make this a reason to request changes, but of course there’s no absolute cut-off in function size before it’s too large to have within a React component so you might disagree, and I’ll just mark as approved if you do.